lib/server.js

```diff
@@ -352,32 +357,32 @@ class Server extends KarmaEventEmitter {
         done(code || 0)
       }
 
-      return this.emitAsync('exit').then(() => {
-        return new Promise((resolve, reject) => {
-          socketServer.sockets.removeAllListeners()
-          socketServer.close()
-          const closeTimeout = setTimeout(removeAllListeners, webServerCloseTimeout)
+      await this.emitAsync('exit')
 
-          webServer.close(() => {
-            clearTimeout(closeTimeout)
-            removeAllListeners()
-            resolve()
-          })
+      return new Promise((resolve) => {
+        socketServer.sockets.removeAllListeners()
+        socketServer.close()
+        const closeTimeout = setTimeout(removeAllListeners, webServerCloseTimeout)
+
+        webServer.close(() => {
+          clearTimeout(closeTimeout)
+          removeAllListeners()
+          resolve()
         })
       })
     }
 
     processWrapper.on('SIGINT', () => disconnectBrowsers(process.exitCode))
     processWrapper.on('SIGTERM', disconnectBrowsers)
 
-    processWrapper.on('uncaughtException', (error) => {
+    processWrapper.on('uncaughtException', async (error) => {
       this.log.error(error)
-      disconnectBrowsers(1)
+      await disconnectBrowsers(1)
```

Comment:
I don't think this is correct and was not correct in the existing code. `disconnectBrowsers()` would previously race the exit calls, the new version just moves the race up the stack one function.  However the race is always won because the exit processing is triggered by the `disconnectBrowsers()` and no work is scheduled here after that call.

I think a comment would do (but without the async) or we could use
https://nodejs.org/api/process.html#process_event_beforeexit
and create a queue of before-exit work.